src/QsCompiler/Compiler/CompilationLoader.cs

```diff
@@ -719,6 +624,90 @@ public CompilationLoader(SourceLoader loadSources, IEnumerable<string> reference
         {
         }
 
+        private IEnumerable<(LoadedStep, Action<Status>)> InternalRewriteSteps(Uri executingAssembly)
+        {
+            // TODO: The dependencies for rewrite steps should be declared as part of IRewriteStep interface, and we
+            // should query those here.
+#pragma warning disable CS0618 // Type or member is obsolete
+            var dependencies = BuiltIn.AllBuiltIns.Select(b => b.FullName);
+#pragma warning restore CS0618 // Type or member is obsolete
+
+            var status = this.compilationStatus;
+            var steps = new (IRewriteStep Step, bool Enabled, Action<Status> SetStatus)[]
+            {
+                // TODO: It would be nicer to trim unused intrinsics. Currently, this is not possible due to how the
+                // old setup of the C# runtime works. With the new setup (interface-based approach for target
+                // packages), it is possible to trim unused intrinsics.
+                (
+                    new SyntaxTreeTrimming(keepAllIntrinsics: true, dependencies),
+                    this.config.IsExecutable && !this.config.SkipSyntaxTreeTrimming,
+                    s => status.TreeTrimming = s),
+
+                (new LiftLambdas(), this.config.LiftLambdaExpressions, s => status.LiftLambdaExpressions = s),
+                (new ClassicallyControlled(), this.config.ConvertClassicalControl, s => status.ConvertClassicalControl = s),
+                (new FunctorGeneration(), this.config.GenerateFunctorSupport, s => status.FunctorSupport = s),
+                (new ConjugationInlining(), !this.config.SkipConjugationInlining, s => status.ConjugationInlining = s),
+                (new FullPreEvaluation(), this.config.AttemptFullPreEvaluation, s => status.PreEvaluation = s),
+                (new Monomorphization(monomorphizeIntrinsics: false), this.config.Monomorphize, s => status.Monomorphization = s),
+                (new CapabilityInference(), !this.config.IsExecutable, s => status.CapabilityInference = s),
+            };
+
+            return from step in steps
+                where step.Enabled
+                select (new LoadedStep(step.Step, typeof(IRewriteStep), executingAssembly), step.SetStatus);
+        }
+
+        private void RunRewriteSteps(Uri executingAssembly)
+        {
+            PerformanceTracking.TaskStart(PerformanceTracking.Task.RewriteSteps);
+            this.config.PrepareQirGeneration = this.config.PrepareQirGeneration || this.externalRewriteSteps.Any(step => step.Name == "QIR Generation");
+
+            var externalSteps = this.externalRewriteSteps.Select((step, index) =>
+                (step, new Action<Status>(status => this.compilationStatus.LoadedRewriteSteps[index] = status)));
+
+            var steps = this.InternalRewriteSteps(executingAssembly)
+                .Concat(externalSteps)
+                .OrderByDescending(step => step.Item1.Priority);

```

Comment:
Before this change, internal rewrite steps were executed before external rewrite steps. I think we should keep the ordering as it was before.

Reply:
Could you explain in more detail? In the old code, it added all rewrite steps (internal and external) to `steps` and sorted them before executing. I think I preserved that behavior.

Reply:
Yes, that's true, I didn't read the code correctly. Resolving this conversation.